File: TPs/TP9/sol.py
# NOTES: translate to cpp

# 1. Build adj list where each neighbor has its cost, excluding 0000
# 2. Calculate distances between 0000 and all keys, then get min_dist (also starting_cost) and starting_key.
# This will be our starting key for Prim. We cant start from 0000 or include it on the graph bc we cant JUMP there.
# If we do it will generate an MST where 0000 could have multiples edges
# 3. Implement Prim but only summing costs (we dont need to save fathers and recreate MST)

def single_key_dist(a, b):
    distance = abs(a - b)
    return min(distance, 10 - distance)

def keys_dist(key1, key2):
    res = 0
    for i in range(4):
        res += single_key_dist(key1[i], key2[i])
    return res

def min_start_cost(keys):
    min_dist = float('inf')
    min_key = (0, 0, 0, 0)
    for key in keys:
        dist = keys_dist((0, 0, 0, 0), key) 
        if dist < min_dist:
            min_key = key
            min_dist = dist
    return min_dist, min_key

def build_graph(keys): # represent graph as adj list with neighbors as a pair<neighbor, weight to that neighbor>
    adj_list = {}
    for key1 in keys:
        neighbors = []
        for key2 in keys:
            if key1 != key2:
                neighbors.append([key2, keys_dist(key1, key2)])
        adj_list[key1] = neighbors
    return adj_list

# Since graph is complete, m ∈ Ω(n^2) => I should use Prim trivial impl O(n^2) rather than heap impl O((m+n)logn) = O(n^2logn). Bc O(n^2) < O(n^2logn)

def prim_min_cost(graph, starting_cost, starting_key):
    total_cost = starting_cost
    visited = [starting_key]

    for _ in range(len(graph)):
        min_dist = float('inf')
        min_key = None
        for key in graph.keys():
            if key not in visited:
                for neighbor in graph[key]:
                    if neighbor[0] in visited:
                        if neighbor[1] < min_dist:
                            min_dist = neighbor[1]
                            min_key = key

        if min_key is None:  
            break

        visited.append(min_key)
        total_cost += min_dist

    return total_cost

T = int(input()) # test cases
keys_global = []

for _ in range(T):
    safe = input().split()
    N = int(safe[0])
    keys = [safe[i] for i in range(1, N+1)]
    keys = [tuple(int(d) for d in key) for key in keys] 
    keys_global.append(keys)

for i in range(T):
    graph = build_graph(keys_global[i])
    starting_cost, starting_key = min_start_cost(keys_global[i])
    min_cost = prim_min_cost(graph, starting_cost, starting_key)
    print(min_cost)

File: TPs/TP9/sol.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdlib>
using namespace std;

// 1. Calculate distances between 0000 and all keys, then get min_dist (starting_cost) and starting_key.
// This will be our starting key for Prim. We cant start from 0000 or include it on the keys we pass to Prim bc we cant JUMP there.
// If we do it will generate an MST where 0000 could have multiples edges.
// 3. Implement Prim only summing costs (we dont need to save fathers and recreate MST) 
// and using dynamic programming for distances (we calculate same distances several times).


vector<string> keys;

int single_key_dist(int a, int b) {
    int distance = abs(a - b);
    return min(distance, 10 - distance);
}

int keys_dist(const string& key1, const string& key2) {
    int distance = 0;
    for (int i=0; i<4; i++) {
        distance += single_key_dist(key1[i] - '0', key2[i] - '0');
    }
    return distance;
}

pair<int, int> min_start_cost(const vector<string>& keys, int N) {
    int min_dist = 26;
    int min_key = -1;
    string init = "0000";
    for (int i=0; i<N; i++) {
        int dist = keys_dist(init, keys[i]);
        if (dist < min_dist) {
            min_key = i;
            min_dist = dist;
        }
    }
    return make_pair(min_dist, min_key);
}

int prim_min_cost(const vector<string>& keys, int starting_cost, int starting_key, int N) {
    vector<int> min_cost(N, 26);
    vector<bool> visited(N, false);
    int total_cost = starting_cost;
    min_cost[starting_key] = 0;

    for (int i=0; i<N; i++) {
        int min_key = -1;
        for (int j=0; j<N; j++) {
            if (!visited[j]) {
                if (min_key == -1 || min_cost[j] < min_cost[min_key]) {
                    min_key = j;
                }
            }
        }
        
        if (min_key == -1) break; 
        
        visited[min_key] = true;
        total_cost += min_cost[min_key];

        for (int k=0; k<N; k++) {
            if (!visited[k]) {
                int dist = keys_dist(keys[min_key], keys[k]);
                if (dist < min_cost[k]) {
                    min_cost[k] = dist;
                }
            }
        }
    }

    return total_cost;
}

int main() {
    int T;
    cin >> T;

    for (int k=0; k<T; k++) {
        int N; cin >> N;
        keys.resize(N);

        for (int i=0; i<N; i++) {
            cin >> keys[i];
        }

        pair<int, int> start = min_start_cost(keys, N);

        int min_cost = prim_min_cost(keys, start.first, start.second, N);

        cout << min_cost << endl;
    }

    return 0;
}
